frontend/src/components/Tree/Tree.jsx

```diff
@@ -0,0 +1,138 @@
+import React, { useState, useRef, useEffect } from "react";
+import styles from "./Tree.module.css";
+import { select, hierarchy, tree, linkHorizontal } from "d3";
+import ResizeObserver from "resize-observer-polyfill";
+import "./Tree.module.css";
+
+const useResizeObserver = (ref) => {
+  const [dimensions, setDimensions] = useState(null);
+  useEffect(() => {
+    const observeTarget = ref.current;
+    const resizeObserver = new ResizeObserver((entries) => {
+      entries.forEach((entry) => {
+        setDimensions(entry.contentRect);
+      });
+    });
+    resizeObserver.observe(observeTarget);
+    return () => {
+      resizeObserver.unobserve(observeTarget);
+    };
+  }, [ref]);
+  return dimensions;
+};
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+export function Tree({ jsonData, onNodeClick }) {
+  const [currentView, setCurrentView] = useState({});
+  const svgRef = useRef();
+  const wrapperRef = useRef();
+  const dimensions = useResizeObserver(wrapperRef);
+
+  // We save data to see if it changed
+  const previouslyRenderedData = usePrevious(currentView);
+
+  useEffect(() => {
+    setCurrentView(jsonData);
```

Comment:
Wondering what the purpose of setCurrentView here is? I was reading that UseEffect() is called after a render, and the state changing results in another render. Therefore when jsonData changes, we would be rendering, then running this effect that causes another render (I noticed this when I used print statements to track the number of renders). Is there any reason why we need currentView given that the onNodeClick is handled elsewhere?

Reply:
Oh thanks for catching that! I was initially doing something else where I needed this variable but now I don't.